src/runtime/sql-refinement-generator.ts

```diff
@@ -0,0 +1,84 @@
+/**
+ * @license
+ * Copyright (c) 2020 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Op} from './manifest-ast-nodes.js';
+import {Dictionary} from './hot.js';
+import {Schema} from './schema.js';
+import {RefinementExpressionVisitor, BinaryExpression, UnaryExpression, FieldNamePrimitive, QueryArgumentPrimitive, BuiltIn, NumberPrimitive, BooleanPrimitive, TextPrimitive, Primitive} from './refiner.js';
+
+const sqlOperator: Dictionary<string> = {
+  [Op.AND]: 'AND',
+  [Op.OR]: 'OR',
+  [Op.LT]: '<',
+  [Op.GT]: '>',
+  [Op.LTE]: '<=',
+  [Op.GTE]: '>=',
+  [Op.ADD]: '+',
+  [Op.SUB]: '-',
+  [Op.MUL]: '*',
+  [Op.DIV]: '/',
+  [Op.NOT]: 'NOT',
+  [Op.NEG]: '-',
+  [Op.EQ]: '=',
+  [Op.NEQ]: '<>',
+};
+
+class SqlRefinementGenerator extends RefinementExpressionVisitor<string> {
+
+  visitBinaryExpression(expr: BinaryExpression): string {
+    return `(${this.visit(expr.leftExpr)} ${sqlOperator[expr.operator.op]} ${this.visit(expr.rightExpr)})`;
+  }
+  visitUnaryExpression(expr: UnaryExpression): string {
+    return `(${sqlOperator[expr.operator.op]} ${this.visit(expr.expr)})`;
+  }
+  visitFieldNamePrimitive(expr: FieldNamePrimitive): string {
+    const fixed = (expr.value.toString());
+    if (expr.evalType === Primitive.BOOLEAN) {
+      return `(${fixed} = 1)`;
+    }
+    return fixed;
+  }
+  visitQueryArgumentPrimitive(expr: QueryArgumentPrimitive): string {
+    return expr.value.toString();
+  }
+  visitBuiltIn(expr: BuiltIn): string {
+    if (expr.value === 'now()') {
+      return expr.value;
+    }
+    // TODO: Implement SQL getter for 'creationTimeStamp'
+    throw new Error(`Unhandled BuiltInNode '${expr.value}' in toSQLExpression`);
+  }
+  visitNumberPrimitive(expr: NumberPrimitive): string {
+    return expr.value.toString();
+  }
+  visitBooleanPrimitive(_: BooleanPrimitive): string {
+    throw new Error('BooleanPrimitive.toSQLExpression should never be called. The expression is assumed to be normalized.');
+  }
+  visitTextPrimitive(expr: TextPrimitive): string {
+    // TODO(cypher1): Consider escaping this for SQL code generation.
+    return `'${expr.value}'`;
```

Comment:
This seems to be a pretty big vector for SQL injection attacks, I think there should be a bug filed to track this.

Reply:
Defnitely. But given we're not generating SQL in our codepaths yet I don't think we should spend time on thinking about security of this class. It is likely to change 5 times before we launch it.

I will put a big comment on top that this is experimental and not prod ready.

Reply:
100% I'm tracking escaping in code generation here in b/159259321 as a security issue to check for in SQL (might be worth checking other code generation as well when we support 3p.